xdsl_smt/utils/transfer_function_util.py

```diff
@@ -0,0 +1,309 @@
+    if len(lst) == 0:
+        assert False and "cannot compress lst with size 0 to an AndOp"
+    elif len(lst) == 1:
+        empty_result: list[Operation] = []
+        return (lst[0].results[0], empty_result)
+    else:
+        new_ops: list[Operation] = [AndOp(lst[0].results[0], lst[1].results[0])]
+        for i in range(2, len(lst)):
+            new_ops.append(AndOp(new_ops[-1].results[0], lst[i].results[0]))
+        return (new_ops[-1].results[0], new_ops)
+
+
+def compareDefiningOp(func: DefineFunOp | None, func1: DefineFunOp | None) -> bool:
+    func_none: bool = func is None
+    func1_none: bool = func1 is None
+    if func_none ^ func1_none:
+        return False
+    if func_none or func1_none:
+        return True
+    for arg, arg1 in zip(func.body.block.args, func1.body.block.args):
+        if arg.type != arg1.type:
+            return False
+    return func.func_type.outputs.data[0] == func1.func_type.outputs.data[0]
+
+
+def fixDefiningOpReturnType(func: DefineFunOp) -> DefineFunOp:
+    smt_func_type = func.func_type
+    ret_val_type = [ret.type for ret in func.return_values]
+    if smt_func_type != ret_val_type:
+        new_smt_func_type = FunctionType.from_lists(
+            smt_func_type.inputs.data, ret_val_type
+        )
+        func.ret.type = new_smt_func_type
+    return func
+
+
+# This class maintains a map from width(int) -> a function
+# When the desired function with given width doesn't exist, it generates one
+# and returns it as the result
+class FunctionCollection:
+    main_func: FuncOp
+    smt_funcs: dict[int, DefineFunOp] = {}
+    create_smt: Callable[[FuncOp, int, MLContext], DefineFunOp]
+    ctx: MLContext
+
+    def __init__(
+        self,
+        func: FuncOp,
+        create_smt: Callable[[FuncOp, int, MLContext], DefineFunOp],
+        ctx: MLContext,
+    ):
+        self.main_func = func
+        self.create_smt = create_smt
+        self.smt_funcs = {}
+        self.ctx = ctx
+
+    def getFunctionByWidth(self, width: int) -> DefineFunOp:
+        if width not in self.smt_funcs:
+            self.smt_funcs[width] = self.create_smt(self.main_func, width, self.ctx)
+        return self.smt_funcs[width]
+
+
+# This class maintains information about a transfer function
+class TransferFunction:
+    is_abstract_arg: list[bool] = []
+    name: str = ""
+    is_forward: bool = True
+    replace_int_attr: bool = False
+    operationNo: int = -1
+    transfer_function: FuncOp
+
+    def __init__(
+        self,
+        transfer_function: FuncOp,
+        is_forward: bool = True,
+        operationNo: int = -1,
+        replace_int_attr: bool = False,
+    ):
+        self.name = transfer_function.sym_name.data
+        self.is_forward = is_forward
+        self.operationNo = operationNo
+        is_abstract_arg: list[bool] = []
+        self.transfer_function = transfer_function
+        func_type = transfer_function.function_type
```

Comment:
```suggestion
        self.transfer_function = transfer_function
        func_type = transfer_function.function_type
        is_abstract_arg: list[bool] = []
```
Moving the variable creation next to its initialization

Reply:
Done!